src/mount-support.c

```diff
@@ -416,6 +241,229 @@ static void sc_setup_mount_profiles(const char *security_tag)
+	debug("performing operation: mount --bind %s %s", dst, dst);
+	if (mount(dst, dst, NULL, MS_BIND, NULL) < 0) {
+		die("cannot perform operation: mount --bind %s %s", dst, dst);
+	}
+	debug("performing operation: mount --make-private %s", dst);
+	if (mount("none", dst, NULL, MS_PRIVATE, NULL) < 0) {
+		die("cannot perform operation: mount --make-private %s", dst);
+	}
+	// On classic mount the nvidia driver. Ideally this would be done in an
+	// uniform way after pivot_root but this is good enough.
+	if (config->on_classic) {
+		sc_mount_nvidia_driver(scratch_dir);
+	}
+	// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+	//                    pivot_root
+	// XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+	// Use pivot_root to "chroot" into the scratch directory.
+	//
+	// Q: Why are we using something as esoteric as pivot_root(2)?
+	// A: Because this makes apparmor handling easy. Using a normal chroot
+	// makes all apparmor rules conditional.  We are either running on an
+	// all-snap system where this would-be chroot didn't happen and all the
+	// rules see / as the root file system _OR_ we are running on top of a
+	// classic distribution and this chroot has now moved all paths to
+	// /tmp/snap.rootfs_*.
+	//
+	// Because we are using unshare(2) with CLONE_NEWNS we can essentially use
+	// pivot_root just like chroot but this makes apparmor unaware of the old
+	// root so everything works okay.
+	debug("performing operation: pivot_root %s %s", scratch_dir, dst);
+	if (syscall(SYS_pivot_root, scratch_dir, dst) < 0) {
+		die("cannot perform operation: pivot_root %s %s", scratch_dir,
+		    dst);
+	}
+	// Unmount the self-bind mount over the scratch directory created earlier
+	// in the original root filesystem (which is now mounted on SC_HOSTFS_DIR).
+	// This way we can remove the temporary directory we created and "clean up"
+	// after ourselves nicely.
+	must_snprintf(dst, sizeof dst, "%s/%s", SC_HOSTFS_DIR, scratch_dir);
+	debug("performing operation: umount %s", dst);
+	if (umount2(dst, 0) < 0) {
+		die("cannot perform operation: umount %s", dst);
+	}
+	// Remove the scratch directory. Note that we are using the path that is
+	// based on the old root filesystem as after pivot_root we cannot guarantee
+	// what is present at the same location normally. (It is probably an empty
+	// /tmp directory that is populated in another place).
+	debug("performing operation: rmdir %s", dst);
+	if (rmdir(scratch_dir) < 0) {
+		die("cannot perform operation: rmdir %s", dst);
+	};
+	// Make the old root filesystem recursively slave. This way operations
+	// performed in this mount namespace will not propagate to the peer group.
+	// This is another essential part of the confinement system.
+	debug("performing operation: mount --make-rslave %s", SC_HOSTFS_DIR);
+	if (mount("none", SC_HOSTFS_DIR, NULL, MS_REC | MS_SLAVE, NULL) < 0) {
+		die("cannot perform operation: mount --make-rslave %s",
+		    SC_HOSTFS_DIR);
+	}
+#if 0
```

Comment:
FYI, I was split on this (advice from stgraber earlier) but it seems to actually do the right thing (to be clear, the code as-is, with `#if 0`). I didn't see any adverse effects either.